sqlglot/generator.py

```diff
@@ -592,10 +592,13 @@ def sql(
         else:
             raise ValueError(f"Expected an Expression. Received {type(expression)}: {expression}")
 
-        sql = self.maybe_comment(sql, expression) if self.comments and comment else sql
+        # We've already generated the comments in prepend_ctes for expressions that have CTEs
+        if self.comments and comment and "with" not in expression.arg_types:
```

Comment:
this doesn't seem to be the most robust way to tell if something has prepend_ctes called on it

Reply:
what if you store a dict of commented things instead

Reply:
Sure I think that works too, this just seemed like a simpler approach but I see why it may not be very robust. I'll do a refactor and implement your suggestion instead.